commandsv3/src/main/java/org/wpilib/commands3/sysid/SysIdRoutine.java

```diff
@@ -0,0 +1,298 @@
+ *
+ * <p>Timestamps are not coordinated across data, so motor frames and test state tags may be
+ * recorded on different log frames. Because frame alignment is not guaranteed, SysId parses the log
+ * by using the test state flag to determine the timestamp range for each section of the test, and
+ * then extracts the motor frames within the valid timestamp ranges. If a given test was run
+ * multiple times in a single logfile, the user will need to select which of the tests to use for
+ * the fit in the analysis tool.
+ */
+public class SysIdRoutine extends SysIdRoutineLog {
+  private final Config m_config;
+  private final SysIdMechanism m_mechanism;
+  private final Consumer<State> m_recordState;
+
+  /**
+   * Create a new SysId characterization routine.
+   *
+   * @param config Hardware-independent parameters for the SysId routine.
+   * @param mechanism Hardware interface for the SysId routine.
+   */
+  public SysIdRoutine(Config config, SysIdMechanism mechanism) {
+    super(mechanism.m_name);
+    m_config = config;
+    m_mechanism = mechanism;
+    m_recordState = config.m_recordState != null ? config.m_recordState : this::recordState;
+  }
+
+  /** Hardware-independent configuration for a SysId test routine. */
+  public static class Config {
+    /** The voltage ramp rate used for quasistatic test routines. */
+    public final Velocity<VoltageUnit> m_rampRate;
+
+    /** The step voltage output used for dynamic test routines. */
+    public final Voltage m_stepVoltage;
+
+    /** Safety timeout for the test routine commands. */
+    public final Time m_timeout;
+
+    /** Optional handle for recording test state in a third-party logging solution. */
+    public final Consumer<State> m_recordState;
+
+    /** Default voltage ramp rate used in quasistatic test routines, set to 1 volt per second. */
+    public static final Velocity<VoltageUnit> DEFAULT_RAMP_RATE = Volts.of(1).per(Second);
+
+    /** Default step voltage used in dynamic test routines, set to 7 volts. */
+    public static final Voltage DEFAULT_STEP_VOLTAGE = Volts.of(7);
+
+    /** Default timeout used to automatically end test routine commands, set to 10 seconds. */
+    public static final Time DEFAULT_TIMEOUT = Seconds.of(10);
+
+    /**
+     * Create a new configuration for a SysId test routine. To use the default config values, pass
+     * in the provided static default fields.
+     *
+     * @param rampRate The voltage ramp rate used for quasistatic test routines.
+     * @param stepVoltage The step voltage output used for dynamic test routines.
+     * @param timeout Safety timeout for the test routine commands.
+     * @param recordState Optional handle for recording test state in a third-party logging
+     *     solution. If provided, the test routine state will be passed to this callback instead of
+     *     logged in WPILog.
+     */
+    public Config(
+        Velocity<VoltageUnit> rampRate,
+        Voltage stepVoltage,
```

Comment:
Is it possible to make these dimension agnostic? For example it would be nice if you could configure a current-based test without having to do some weird trickery.

Reply:
As in have them accept doubles? The ideal way would really be a type union of Voltage and Current but doing that in Java is very verbose so IDK.

Reply:
That could be useful as a followup but the unit handling behind the scenes is already kind of messy (in the sense of converting everything to a certain set of units before recording them regardless of what is passed in) so id prefer to save this for when that is cleaned up, which depends on a proper NT metadata spec.